tests/loops/test_evaluation_loop.py

```diff
@@ -42,6 +44,28 @@ def test_on_evaluation_epoch_end(eval_epoch_end_mock, tmpdir):
     assert eval_epoch_end_mock.call_count == 4
 
 
+def test_set_epoch_called_eval_predict(tmpdir):
+    """Tests that set_epoch (if the sampler has one) is called on the DataLoader during evaluation and
+    prediction."""
+    model = BoringModel()
+    dataset = RandomDataset(32, 64)
+    sampler = RandomSampler(dataset)
+    sampler.set_epoch = Mock()
+    dataloader = torch.utils.data.dataloader.DataLoader(dataset, sampler=sampler)  # DataLoader
+    trainer = Trainer(
+        default_root_dir=tmpdir, limit_train_batches=2, limit_val_batches=2, max_epochs=2, enable_model_summary=False
+    )
+
+    trainer.fit(model, dataloader)
+    # One for each epoch!
+    assert dataloader.sampler.set_epoch.call_count == 2
+
+    predict_dataloader = torch.utils.data.dataloader.DataLoader(dataset, sampler=sampler)
+    trainer.predict(model, predict_dataloader)
+    # One for each epoch!
+    assert predict_dataloader.sampler.set_epoch.call_count == 2
```

Comment:
This only returns 2 because of the previous value that the mock is holding from fit(). 
If you `sampler.reset_mock()` then it returns 0. 
This is because the trainer replaces the sampler no matter what the user sets there (shuffle=True or sampler=). 

I looked at the linked issue again and there the user wants to shuffle the data for validation and test. This is currently not possible and setting the epoch for the distributed sampler won't matter, because shuffling is turned off. 

The issue is actually directly related to #11856. So without directly resolving a decision for #11856, setting the epoch here won't matter. Sorry that I didn't spot this relation earlier. I should have noticed.

Reply:
>  This is currently not possible and setting the epoch for the distributed sampler won't matter, because shuffling is turned off.

for val/test/predict, shuffle will depend upon the dataloader passed and will be used will initializing the `DistributedSampler`.